### PR TITLE
6418 zpool should have a label clearing command

### PR DESCRIPTION
**Pull request by yuripv**
A follow-up work on https://github.com/openzfs/openzfs/pull/32

### CODE REVIEW

**Comment by ahrens** [comment]
@zettabot go

**Comment by yuripankov** [comment]
@ahrens sorry, I didn't notice you already started the tests and pushed again (fixing several typos), could we restart the tests please?

**Comment by ahrens** [comment]
@zettabot go

File: usr/src/lib/libzfs/common/libzfs_import.c [hunk 1]
```diff
@@ -1174,8 +1172,8 @@ zpool_find_import_impl(libzfs_handle_t *hdl, importargs_t *iarg)
 		 * reading the labels skips a bunch of slow operations during
 		 * close(2) processing, so we replace /dev/dsk with /dev/rdsk.
 		 */
-		if (strcmp(path, "/dev/dsk/") == 0)
-			rdsk = "/dev/rdsk/";
+		if (strcmp(path, ZFS_DISK_ROOTD) == 0)
+			(void) strncpy(path, ZFS_RDISK_ROOTD, sizeof (path));
```
Comment (yuripankov) [inline usr/src/lib/libzfs/common/libzfs_import.c]: Obvious bug here, copying to wrong variable, will fix shortly.

File: usr/src/lib/libzfs/common/libzfs_pool.c [hunk 2]
```diff
@@ -3350,7 +3376,7 @@ zpool_vdev_name(libzfs_handle_t *hdl, zpool_handle_t *zhp, nvlist_t *nv,
 				devid_str_free(newdevid);
 		}
 
-		if (strncmp(path, "/dev/dsk/", 9) == 0)
+		if (strncmp(path, ZFS_DISK_ROOT, 9) == 0)
```
Comment (yuripankov) [inline usr/src/lib/libzfs/common/libzfs_pool.c]: Obvious bug here, should be ZFS_DISK_ROOTD.

**Comment by yuripankov** [comment]
After re-reading the discussion in original PR, especially what @ilovezfs said, I've made the command use the following logic:

- if provided vdev name contains absolute path, use it as is
- check if it points to existing file and use it as is
- prepend /dev/dsk and append s0 (if needed, I understand that it's not portable but that's the best I could come up with) and check if it points to existing device
- error out if previous steps didn't point us to existing vdev

- try opening the device, error out on failure

- try reading the label, error out if it's not there

- check if vdev is part of active (error out) or exported/foreign (require -f) pool

- finally, wipe the label

Fix several obvious bugs from the original change while here.

Example: https://gist.github.com/yuripankov/f2f728c5d32cfadacc7ba91515791768

**Comment by ahrens** [comment]
@zettabot go

File: usr/src/cmd/zpool/zpool_main.c [hunk 3]
```diff
@@ -618,6 +624,141 @@ zpool_do_remove(int argc, char **argv)
...
+			    optopt);
+			usage(B_FALSE);
+		}
+	}
+
+	argc -= optind;
+	argv += optind;
+
+	/* get vdev name */
+	if (argc < 1) {
```
Comment (ahrens) [inline usr/src/cmd/zpool/zpool_main.c]: What if argc > 1 ?  looks like we silently ignore additional arguments.  We should make that an error (or, allow them to specify a list of vdevs and clear all their labels).

File: usr/src/cmd/zpool/zpool_main.c [hunk 4]
```diff
@@ -618,6 +624,141 @@ zpool_do_remove(int argc, char **argv)
...
+	/*
+	 * Check if we were given absolute path and use it as is.
+	 * Otherwise if the provided vdev name doesn't point to a file,
+	 * try prepending dsk path and appending s0.
+	 */
+	(void) strlcpy(vdev, argv[0], sizeof (vdev));
+	if (vdev[0] != '/' && stat(vdev, &st) != 0) {
+		(void) snprintf(vdev, sizeof (vdev), "%s/%s",
+		    ZFS_DISK_ROOT, argv[0]);
+		if (strstr(vdev, "s0") == NULL)
```
Comment (ahrens) [inline usr/src/cmd/zpool/zpool_main.c]: This is going to be annoying if you specify `c0t1d2s3` (which you could have specified to `zpool add`.  I guess the only good alternative would be to only append `s0` only if there isn't a "s<n>" at the end.  But this (non-zero slice) is an unusual use case so IMO it isn't that bad for them to have to do `/dev/dsk/c1t1d0s3` in this case.

File: usr/src/cmd/zpool/zpool_main.c [hunk 5]
```diff
@@ -618,6 +624,141 @@ zpool_do_remove(int argc, char **argv)
...
+			(void) strlcat(vdev, "s0", sizeof (vdev));
+		if (stat(vdev, &st) != 0) {
+			(void) fprintf(stderr, gettext(
+			    "failed to find device name for %s\n"), argv[0]);
```
Comment (ahrens) [inline usr/src/cmd/zpool/zpool_main.c]: How about "failed to find device '%s', try specifying absolute path instead"

File: usr/src/cmd/zpool/zpool_main.c [hunk 6]
```diff
@@ -618,6 +624,141 @@ zpool_do_remove(int argc, char **argv)
...
+
+	ret = zpool_in_use(g_zfs, fd, &state, &name, &inuse);
+	if (ret != 0) {
+		(void) fprintf(stderr,
+		    gettext("failed to check state for %s\n"), vdev);
+		return (1);
+	}
+
+	if (!inuse)
+		goto wipe_label;
```
Comment (ahrens) [inline usr/src/cmd/zpool/zpool_main.c]: We could avoid the "goto" by doing
```if (inuse) {
  switch (state) {
    ...
```
at the cost of another indent and maybe making the strings even harder to read.  I'm OK either way, up to you.

Comment (yuripankov) [inline usr/src/cmd/zpool/zpool_main.c]: It was that way originally, I added the goto to reduce indentation.

File: usr/src/cmd/zpool/zpool_main.c (hunk 5 above)
Comment (yuripankov) [inline usr/src/cmd/zpool/zpool_main.c]: OK.

File: usr/src/cmd/zpool/zpool_main.c (hunk 4 above)
Comment (yuripankov) [inline usr/src/cmd/zpool/zpool_main.c]: Will fix.

File: usr/src/cmd/zpool/zpool_main.c (hunk 3 above)
Comment (yuripankov) [inline usr/src/cmd/zpool/zpool_main.c]: Allowing to specify more than one vdev will make it needlessly complicated. I'll simply make it return error if argc > 1.

File: usr/src/lib/libzfs/common/libzfs_import.c [hunk 7]
```diff
@@ -1174,10 +1172,10 @@ zpool_find_import_impl(libzfs_handle_t *hdl, importargs_t *iarg)
 		 * reading the labels skips a bunch of slow operations during
 		 * close(2) processing, so we replace /dev/dsk with /dev/rdsk.
 		 */
-		if (strcmp(path, "/dev/dsk/") == 0)
-			rdsk = "/dev/rdsk/";
+		if (strcmp(path, ZFS_DISK_ROOTD) == 0)
+			(void) strlcpy(rdsk, ZFS_RDISK_ROOTD, sizeof (rdsk));
 		else
-			rdsk = path;
+			(void) strlcpy(rdsk, path, sizeof(rdsk));
```
Comment (ahrens) [inline usr/src/lib/libzfs/common/libzfs_import.c]: I think it's supposed to be `sizeof (rdsk)` (with a space) -- or at least the same as 2 lines above.

Comment (yuripankov) [inline usr/src/lib/libzfs/common/libzfs_import.c]: Right... Will fix.

**Comment by ahrens** [comment]
Thanks a lot for picking this up, @yuripankov! These changes look really good.  The only other thing I could ask for is to add tests for this to the test suite (helps ensure that others don't accidentally break this in the future).  But I would be happy to take this change without tests too.

**Comment by yuripankov** [comment]
Updated with the changes you pointed out.

**Comment by ilovezfs** [comment]
@yuripankov Thank you for working on this!

Should there been an override for "try reading the label, error out if it's not there"?

**Comment by yuripankov** [comment]
Not at the moment, though I thought about that, may be when we have absolute path specified, -ff should override all checks and simply do the label clear work?

**Comment by ilovezfs** [comment]
I think something like that (-ff? -F? -fF? --really-I-meant-it?) might be desirable in the event that we're trying to clear the label precisely because it cannot be parsed successfully and is somehow damaged, or possibly because it's being scripted and should be re-runnable, etc.

**Comment by yuripankov** [comment]
Added, looks like a really nice way to break your pool :)

antares:root:~# zpool create data raidz c4t0d0 c4t1d0 c4t2d0
antares:root:~# zpool labelclear c4t0d0
/dev/dsk/c4t0d0s0 is a member (ACTIVE) of pool "data"
antares:root:~# zpool labelclear /dev/dsk/c4t0d0
failed to read label from /dev/dsk/c4t0d0
antares:root:~# zpool labelclear -f /dev/dsk/c4t0d0
failed to read label from /dev/dsk/c4t0d0
antares:root:~# zpool labelclear -ff /dev/dsk/c4t0d0
antares:root:~# zpool status data
  pool: data
 state: ONLINE
status: One or more devices could not be used because the label is missing or
        invalid.  Sufficient replicas exist for the pool to continue
        functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: http://illumos.org/msg/ZFS-8000-4J
  scan: scrub repaired 0 in 0h0m with 0 errors on Sun Apr  3 05:08:38 2016
config:

        NAME        STATE     READ WRITE CKSUM
        data        ONLINE       0     0     0
          raidz1-0  ONLINE       0     0     0
            c4t0d0  UNAVAIL      0     0     0  corrupted data
            c4t1d0  ONLINE       0     0     0
            c4t2d0  ONLINE       0     0     0

errors: No known data errors
antares:root:~#

**Comment by ilovezfs** [comment]
@yuripankov Yes! And combined with a partitioning tool you could probably turn it into a `dd if=/dev/zero ...` replacement and zero the entire device! :)
(This is an exercise left to the reader.)

**Comment by ahrens** [comment]
I have 2 concerns with `-ff`:
1. I think the right flag would be `-F`, which is more in keeping with the existing CLI (e.g. see `zpool import -f` / `-F`, and `zfs receive -F`)
2. I'm not a big fan of making `zpool` overwrite unknown data. Even `zpool create` / `zpool add` will give a specific warning if it detects another filesystem type (e.g. UFS) present on the device, and it will refuse to add the device if it's currently mounted.  Can we make `zpool labelclear -F` at least that conservative?

```
# newfs /dev/dsk/c1t1d0s0
# zpool create test c1t1d0s0
invalid vdev specification
use '-f' to override the following errors:
/dev/dsk/c1t1d0s0 contains a ufs filesystem.
# zpool create test c1t1d0
invalid vdev specification
use '-f' to override the following errors:
/dev/dsk/c1t1d0s0 contains a ufs filesystem.
# mount /dev/dsk/c1t1d0s0 /mnt
# zpool create -f test c1t1d0
invalid vdev specification
the following errors must be manually repaired:
/dev/dsk/c1t1d0s0 is currently mounted on /mnt. Please see umount(1M).
# zpool create -f test c1t1d0
<success>
```

**Comment by yuripankov** [comment]
Well, my thought (and I guess same for @ilovezfs) was that -ff (and that's exactly why it isn't -F) should be treated as "I really know what I'm doing, just do it", without any checks. We could skip documenting it, but I see the use for it in, e.g., installers. If you really don't like it, I can remove last commit and just go ahead with finally integrating what was reviewed previously.

**Comment by ahrens** [comment]
@yuripankov @ilovezfs What would be the use case for "there is no ZFS label but I want to zero out the place that it belongs" (i.e. `zpool labelclear -ff`)? If there's a compelling use case then it's worth considering.  Otherwise I think we should leave it out.

**Comment by yuripankov** [comment]
@ahrens I've removed the last commit, let's integrate it as it is now.

**Comment by ahrens** [comment]
@zettabot go

**Comment by ahrens** [comment]
Changes look good to me.  Thanks for finishing this, @yuripankov.
@grwilson can you take a look at this too?

**Comment by ilovezfs** [comment]
@ahrens The use case is that the "there is no ZFS label" code is returning a false negative, I know it's returning a false negative, and I know what I want to do.

**Comment by ilovezfs** [comment]
The intent of the labelclear command is often to handle cases where there is a problem with the parsing of the label. So blocking the command entirely when there is a problem with the parsing of the label that causes ZFS to conclude "there is no label" makes the command significantly less useful.

**Comment by ahrens** [comment]
@ilovezfs It sounds like you're saying that in some circumstance the label is both there and not there.  What code is detecting a label when `zpool_read_label()` does not find a label?  Maybe we should have that code and `zpool_read_label()` have the same criteria for detecting a label.

Do you often get into this situation?  Do you have any idea how it happens?

**Comment by ilovezfs** [comment]
@ahrens I think the most common ways it happens is when there's a confusion with the whole device vs. a partition, or where there are coincidental overlaps between where a label would be expected with two different partition schemes (e.g., especially when a partition extends to the end of the device and does not leave space at the end), or when the partition table has been changed or has just been re-created during the zpool create/replace/attach/etc. There also discrepancies between when one platform thinks there's a valid label and another platform doesn't and there shouldn't be cases where labelclear is happy to proceed on one platform, and not happy to proceed on another, since then we end up with the goofy, but common, debugging advice "well you could try booting FreeBSD to see if it works there."

My main point though is that the command should have an option that makes it idempotent. Presumably I want to run the command because I want a device without a label at the end of the command, not because I want it second guessing me. If I run it once or ten times, my goal is the same. When I run the command and there is already a label there, my goal is the same as when I run the command and there is not already a label there. Not having -F/-ff/whatever is like having the rm command without the -f option, the mkdir command without the -p option, a touch command that requires the file to either already exist or not to already exist, zfs clone without -p, etc. For scripting purposes it is always better to at least have the option of an idempotent version of a command than to not have it at all.

If I wanted to know ZFS's opinion about whether there's a label there, I'd use zdb -l not zpool labelclear.

A: "Please clear the signature field so I can resign."
B: "But, sir, we cannot clear the signature field because that's not a valid signature."
A: "Yes, I know, I want to sign again so it is a valid signature."
B: "We're sorry, sir, but we can only clear signature fields containing valid signatures."

Deliberately setting up the possibility of Catch 22 scenarios where the recourse is "well you could use dd" seems counter-productive.

If ZFS's parsing of labels always made sense, and always reliably determined when there's a label or not, why do we need this command at all? zpool create/replace/attach/online/offline/etc., especially with the availability of -f options, should always just work.

**Comment by richardelling** [comment]
> On Apr 4, 2016, at 10:48 PM, ilovezfs <notifications@github.com> wrote:
> 
> @ahrens <https://github.com/ahrens> I think the most common ways it happens is when there's a confusion with the whole device vs. a partition, or where there are coincidental overlaps between where a label would be expected with two different partition schemes (e.g., especially when a partition extends to the end of the device and does not leave space at the end), or when the partition table has been changed or has just been re-created during the zpool create/replace/attach/etc. There also discrepancies between when one platform thinks there's a valid label and another platform doesn't and there shouldn't be cases where labelclear is happy to proceed on one platform, and not happy to proceed on another, since then we end up with the goofy, but common, debugging advice "well you could try booting FreeBSD to see if it works there."
> 
> My main point though is that the command should have an option that makes it idempotent. Presumably I want to run the command because I want a device without a label at the end of the command, not because I want it second guessing me. If I run it once or ten times, my goal is the same. When I run the command and there is already a label there, my goal is the same as when I run the command and there is not already a label there. Not having -F/-ff/whatever is like having the rm command without the -f option, the mkdir command without the -p option, a touch command that requires the file to either already exist or not to already exist, zfs clone without -p, etc. For scripting purposes it is always better to at least have the option of an idempotent version of a command than to not have it at all.
> 
> If I wanted to know ZFS's opinion about whether there's a label there, I'd use zdb -l not zpool labelclear.
> 
> A: "Please clear the signature field so I can resign."
> B: "But, sir, we cannot clear the signature field because that's not a valid signature."
> A: "Yes, I know, I want to sign again so it is a valid signature."
> B: "We're sorry, sir, but we can only clear signature fields containing valid signatures."
> 
> Deliberately setting up the possibility of Catch 22 scenarios where the recourse is "well you could use dd" seems counter-productive.
> 
> 

IMNSHO, if you really intend to destroy data, then use dd. ZFS should not be used to destroy data
willy-nilly.
 -- richard

**Comment by ilovezfs** [comment]
@richardelling Would you object to not having the command at all?

**Comment by ptx0** [comment]
I don't think dd is the alternative to having zpool perform the operation you request, even if it is insane. how about the `-f` flag ?

**Comment by richardelling** [comment]
> On Apr 5, 2016, at 8:24 AM, ilovezfs <notifications@github.com> wrote:
> 
> @richardelling <https://github.com/richardelling> Would you object to not having the command at all?
> 
> 

The command is useful for developers, no question. But it is a loaded gun
and adding more powerful bullets doesn't improve the protection.
 -- richard

**Comment by ahrens** [comment]
@ilovezfs I would put the analogy this way:
> A: "Please clear the signature field so I can resign."
B: That doesn't look like a signature field, so I'm not going to clear it.  [optional: In fact, it looks like this space is being used for other purpose X]

> there shouldn't be cases where labelclear is happy to proceed on one platform, and not happy to proceed on another
I agree, and we can address this by making `zpool labelclear -f` work on anything that can be considered a ZFS label by any platform.  E.g. if some platforms don't write "state" and "txg" to the label, we could make a "super relaxed mode" of `zpool_read_label` that doesn't require those fields in the config?

**Comment by ilovezfs** [comment]
I value a more useful CLI with a re-runnable *option* over saving users from themselves, but I guess that's selfish. While we're at it, `zfs receive -F` should probably go the way of the dodo since it has an undocumented ability to destroy datasets, not just roll them back.

**Comment by ahrens** [comment]
@ilovezfs The difference is that `zfs receive -F` won't break anything -- it won't result in inconsistent filesystems.  Versus `zpool labelclear -F` doesn't know what it's overwriting so it could arbitrarily break things (e.g. overwrite a mounted UFS filesystem).  That said, I will reconsider if you have a specific example of when it would be required.

Simply put, `zpool labelclear` removes the zfs labels from the disk.  This is not exactly the same as overwriting the places that the zfs label can be stored on disk.

> zfs receive -F ... has an undocumented ability to destroy datasets

The behavior of `zfs receive -F` is documented in the zfs(1m) manpage:

           -F
               Force a rollback of the file system to the most recent snapshot
               before performing the receive operation. If receiving an
               incremental replication stream (for example, one generated by
               zfs send -R -[iI]), destroy snapshots and file systems that do
               not exist on the sending side.

**Comment by ilovezfs** [comment]
Nope, go ahead.

As for `receive -F` I'm referring to the way a non-incremental `receive -F` will blow away an existing file system of the same name even if it has no snapshots, which cannot be considered a rollback.

While we're on the subject of overwriting things when we don't what we're overwriting and thereby breaking things, it would be awesome if https://www.illumos.org/issues/6477 could get some attention because it's a *very* dangerous issue on OS X (and potentially Linux as well for that matter).

**Comment by ahrens** [comment]
@ilovezfs Ah, I see.  Technically, it is equivalent to a rollback (to the empty $ORIGIN@$ORIGIN), but I agree that this kind of "blow away" receive might be surprising.  At a minimum, we should document this behavior better.

I agree that https://www.illumos.org/issues/6477 is dangerous!  I'll see if I can find some time to investigate.

**Comment by ilovezfs** [comment]
@ahrens Yup, I figured under the hood the destruction was "actually" a rollback in the most trivial sense, but it certainly wouldn't be seen as such by the victim. :)

Did I mention that I can destroy arbitrary, unrelated HFS+ file systems using https://www.illumos.org/issues/6477 simply by running an "unfortunate" zpool import? Anyway, that already keeps me up at night more than a *documented* labelclear -F ever would.

**Comment by yuripankov** [comment]
I've pushed an update, fixing small bug I introduced while addressing comments. Can we please get this tested and integrated?

**Comment by ahrens** [comment]
@zettabot go

**Comment by ahrens** [comment]
@yuripankov Thanks for your work on this.  I'll take a final pass this weekend and get it committed.  Do you want to be listed as the author (vs Will) or code reviewer?

Any other reviewers?  @ilovezfs?

**Comment by yuripankov** [comment]
I think @wca should be the author, and I should be reviewer.

**Comment by yuripankov** [comment]
Fixed a small nit - use argv[0], not vdev, when looking for the sN - this doesn't change the behavior, but looks more correct.

**Comment by ahrens** [comment]
@zettabot go

**Pull request closed without merging**